### ISSUE TITLE
Declarative Agent mountPoints and portMappings are broken and need to be documented

### ISSUE THREAD

**Issue by jacob-ablowitz** [jenkinsci/amazon-ecs-plugin#160]
I am using the ECS plugin to issue build tasks, and ports that should be exposed are not. Agent definition:

```
<... other pipeline elements ...>
    stage('Testing starts after this point...') {
      agent { 
        ecs {
          cloud '<cloud string>'
          label '<label string>'
          image "<tag string>:${env.BUILD_TAG}"
          launchType 'FARGATE'
          remoteFSRoot '/home/jenkins'
          memoryReservation 2048
          cpu 1024
          subnets '<subnet id>'
          securityGroups '<sg id>'
          assignPublicIp true
          portMappings([[containerPort: 8080, hostPort: 8080, protocol: 'tcp']])
          sharedMemorySize 0
          taskrole '<arn details>:role/ecsTaskExecutionRole'
          executionRole '<arn details>:role/ecsTaskExecutionRole' 
        }
      }
   <...other stage items using that agent...>
    }
  <...other pipeline elements...>
```

A sample of logs in `/var/log/jenkins/jenkins.log` after running this pipeline:

```
2020-03-24 00:09:11.894+0000 [id=123]   INFO    o.j.p.g.w.s.DefaultPushGHEventSubscriber#onEvent: Received PushEvent for https://github.com/Valid-Eval/ve_new_backend from 140.82.115.245 ⇒ https://jenkins.valideval.com/github-webhook/
2020-03-24 00:09:27.791+0000 [id=6750]  INFO    c.c.j.p.a.p.ECSDeclarativeAgent#getAsArgs: In getAsArgs. label: jenkins-ecs-backend-agent
2020-03-24 00:09:27.791+0000 [id=6750]  INFO    c.c.j.p.a.p.ECSDeclarativeAgent#getAsArgs: In getAsArgs. argMap: {assign PublicIp=true, cloud=ve-jenkins-build-cloud, cpu=1024, executionRole=arn:aws:iam::030282970159:role/ecsTaskExecutionRole, image=030282970159.dkr.ecr.us-east-1.amazonaws.com/ve-backend-cicd:jenkins-ValidEval-ve_new_backend-jenkins-test-159, label=jenkins-ecs-backend-agent, launchType=FARGATE, memoryReservation=2048, name=jenkins-ecs-backend-agent, overrides=[assignPublicIp, cloud, cpu, executionRole, image, label, launchType, memoryReservation, portMappings, remoteFSRoot, securityGroups, sharedMemorySize, subnets, taskrole], privileged=false, remoteFSRoot=/home/jenkins, securityGroups=sg-00f4febf613f2b171, subnets=subnet-99f3c7c5, taskrole=arn:aws:iam::030282970159:role/ecsTaskExecutionRole, uniqueRemoteFSRoot=false}
2020-03-24 00:09:27.799+0000 [id=6750]  INFO    c.c.j.p.a.p.ECSTaskTemplateStepExecution#start: Registering task template with name jenkins-ecs-backend-agent-4z4l2
2020-03-24 00:09:32.523+0000 [id=6791]  INFO    h.TcpSlaveAgentListener$ConnectionHandler#run: Connection #828 failed: java.io.EOFException
2020-03-24 00:09:34.600+0000 [id=6796]  INFO    h.TcpSlaveAgentListener$ConnectionHandler#run: Connection #829 failed: java.io.EOFException
2020-03-24 00:09:37.235+0000 [id=34]    INFO    c.c.j.plugins.amazonecs.ECSCloud#provision: Asked to provision 1 agent(s) for: jenkins-ecs-backend-agent
2020-03-24 00:09:37.235+0000 [id=34]    INFO    c.c.j.plugins.amazonecs.ECSCloud#provision: Will provision ve-jenkins-build-cloud-jenkins-ecs-backend-agent-zbpx3, for label: jenkins-ecs-backend-agent
2020-03-24 00:09:47.242+0000 [id=39]    INFO    hudson.slaves.NodeProvisioner#lambda$update$6: ve-jenkins-build-cloud-jenkins-ecs-backend-agent-zbpx3 provisioning successfully completed. We have now 3 computer(s)
2020-03-24 00:09:47.286+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSService#findTaskDefinition: No existing task definition found for family or ARN: ve-jenkins-build-cloud-jenkins-ecs-backend-agent-4z4l2
2020-03-24 00:09:47.351+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSService#registerTemplate: Created Task Definition: {TaskDefinitionArn: <arn id>:task-definition/<agent name>-4z4l2:1,ContainerDefinitions: [{Name: <agent name>-4z4l2,Image: <image id>,Cpu: 1024,MemoryReservation: 2048,Links: [],PortMappings: [],Essential: true,EntryPoint: [],Command: [],Environment: [],MountPoints: [],VolumesFrom: [],Secrets: [],DependsOn: [],Privileged: false,DnsServers: [],DnsSearchDomains: [],ExtraHosts: [],DockerSecurityOptions: [],Ulimits: [],SystemControls: [],ResourceRequirements: [],}],Family: <agent name>-4z4l2,TaskRoleArn: <arn id>:role/ecsTaskExecutionRole,ExecutionRoleArn: <arn id>:role/ecsTaskExecutionRole,NetworkMode: awsvpc,Revision: 1,Volumes: [],Status: ACTIVE,RequiresAttributes: [{Name: com.amazonaws.ecs.capability.ecr-auth,}, {Name: com.amazonaws.ecs.capability.docker-remote-api.1.21,}, {Name: com.amazonaws.ecs.capability.task-iam-role,}, {Name: ecs.capability.execution-role-ecr-pull,}, {Name: com.amazonaws.ecs.capability.docker-remote-api.1.18,}, {Name: ecs.capability.task-eni,}],PlacementConstraints: [],Compatibilities: [EC2, FARGATE],RequiresCompatibilities: [FARGATE],Cpu: 1024,Memory: 2048,InferenceAccelerators: [],}
2020-03-24 00:09:47.351+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#runECSTask: [<agent name>-zbpx3]: Starting agent with task definition <arn id>:task-definition/<agent name>-4z4l2:1}
2020-03-24 00:09:48.088+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#runECSTask: [<agent name>-zbpx3]: Agent started with task arn : <arn id>:task/1498419b-daff-4c67-84b4-4e726adf14c5
2020-03-24 00:09:48.089+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#launch: [<agent name>-zbpx3]: TaskArn: <arn id>:task/1498419b-daff-4c67-84b4-4e726adf14c5
2020-03-24 00:09:48.089+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#launch: [<agent name>-zbpx3]: TaskDefinitionArn: <arn id>:task-definition/<agent name>-4z4l2:1
2020-03-24 00:09:48.089+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#launch: [<agent name>-zbpx3]: ClusterArn: <arn id>:cluster/<cluster name>
2020-03-24 00:09:48.089+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#launch: [<agent name>-zbpx3]: ContainerInstanceArn: null
2020-03-24 00:09:48.129+0000 [id=6714]  INFO    c.c.j.p.amazonecs.ECSLauncher#launch: [<agent name>-zbpx3]: Waiting for agent to start
```
and it continues with "Waiting for agent to start" messages until the agent starts and runs in the defined AWS ECS task.

Note that in the above, `<something>` in angle brackets is my addition and intended to denote a properly-formatted item that appears to be working correctly / reflected in the logs correctly.

From the logs it appears as though the absence of `portMappings` in the `ECSDeclarativeAgent#getAsArgs` log entry and `PortMappings: []` in the `ECSService#registerTemplate` log entry suggests no port mappings are being sent to AWS, and the behavior of the container and the task logs in AWS are consistent with the port not being open.

I originally believed my syntax was likely incorrect, because the use of the `portMappings` option is not documented and I had inferred settings from UI HTML inspection, but after additional experimentation and troubleshooting realized that was not the case (i.e., my syntax was fine). So I investigated further and detected incorrect null variable checks for `mountPoints` and `portMappings` in `ECSDeclarativeAgent#getAsArgs`, likely the result of a copy-paste from implementing `extraHosts`. I will submit a PR shortly from a forked version of the repo which addresses both issues. Your quick attention to this PR is much appreciated...

**Comment by jacob-dmi** [comment]
Added https://github.com/jenkinsci/amazon-ecs-plugin/pull/161

**Comment by webratz** [comment]
#161 is merged

**Comment by webratz** [comment]
can you confirm its fixed with #161 (released now as 1.27)

**Comment by jacob-ablowitz** [comment]
Sorry, never saw the request to confirm. Yes this is fixed with #161 and this issue should be marked resolved.
